### PR TITLE
Add heatmap overlays on Android and iOS

### PR DESCRIPTION
**Pull request by Zwatotem**

### CODE REVIEW

File: expo-maps/android/src/main/java/expo/modules/maps/PropsData.kt [hunk 1]
```diff
@@ -39,6 +40,11 @@ data class MarkerObject(
 }
 
 data class Point(@Field val latitude: Double = 0.0, @Field val longitude: Double = 0.0) : Record
+data class PointWithData(
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/PropsData.kt:47]:
```suggestion

data class PointWithData(
```

File: expo-maps/android/src/main/java/expo/modules/maps/PropsData.kt [hunk 2]
```diff
@@ -107,3 +113,33 @@ data class ClusterObject(
...
+    return result
+  }
+}
+
+data class HeatmapObject(
+  @Field val points: List<PointWithData> = emptyList(),
+  @Field val radius: Int?,
+  @Field val gradient: Gradient? = null,
+  @Field val opacity: Double?,
+) : Record
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/PropsData.kt:179]:
```suggestion
) : Record

```

File: expo-maps/android/src/main/java/expo/modules/maps/googleMaps/GoogleMapsHeatmaps.kt [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+class GoogleMapsHeatmaps(private val context: Context, private val map: GoogleMap) : Heatmaps {
+
+  private val heatmapOverlays = mutableListOf<TileOverlay>()
+
+  override fun setHeatmaps(heatmapObjects: Array<HeatmapObject>) {
+    heatmapOverlays.forEach { it.remove() }
+    heatmapOverlays.clear()
+
+    for(it in heatmapObjects) {
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/googleMaps/GoogleMapsHeatmaps.kt]:
```suggestion
    for(heatmapObject in heatmapObjects) {
```

File: expo-maps/android/src/main/java/expo/modules/maps/googleMaps/GoogleMapsHeatmaps.kt [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        Integer.decode("0x" + colorString.substring(5..6)),
+      )
+      9 -> Color.argb(
+        Integer.decode("0x" + colorString.substring(7..8)),
+        Integer.decode("0x" + colorString.substring(1..2)),
+        Integer.decode("0x" + colorString.substring(3..4)),
+        Integer.decode("0x" + colorString.substring(5..6)),
+      )
+      else -> throw IllegalArgumentException("String $colorString is not a valid color representation")
+    }
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/googleMaps/GoogleMapsHeatmaps.kt]: After you merge you should have same function named `colorHexStringToInt` available in `GoogleMapsUtils.kt` file, so you can reuse it to not duplicate code.

File: expo-maps/android/src/main/java/expo/modules/maps/interfaces/Heatmaps.kt [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package expo.modules.maps.interfaces
+
+import expo.modules.maps.HeatmapObject
+
+interface Heatmaps {
+  fun setHeatmaps(heatmapObjects: Array<HeatmapObject>)
+}
```
Comment (szdziedzic) [inline expo-maps/android/src/main/java/expo/modules/maps/interfaces/Heatmaps.kt]:
```suggestion
}

```

**Review by szdziedzic — COMMENTED** [review]
Nice!

File: expo-maps/android/src/main/java/expo/modules/maps/PropsData.kt [hunk 6]
```diff
@@ -107,3 +113,33 @@ data class ClusterObject(
...
+    if (!locations.contentEquals(other.locations)) return false
+
+    return true
+  }
+
+  override fun hashCode(): Int {
+    var result = colors.hashCode()
+    result = 31 * result + locations.contentHashCode()
+    return result
+  }
```
Comment (p-karcz) [inline expo-maps/android/src/main/java/expo/modules/maps/PropsData.kt:171]: No need for it, data class automatically overrides these with default implementation that compares instances by looking on members' values, hashcode is overridden too

File: expo-maps/android/src/main/java/expo/modules/maps/googleMaps/GoogleMapsHeatmaps.kt [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+      }
+      it.radius.let {
+        builder = builder.radius(it ?: 20)
+      }
+      it.opacity.let {
+        builder = builder.opacity(it ?: 1.0)
+      }
+      var provider = builder.build()
+      val tileOverlay = map.addTileOverlay(TileOverlayOptions().tileProvider(provider))
+      tileOverlay.let { heatmapOverlays.add(it!!) }
```
Comment (p-karcz) [inline expo-maps/android/src/main/java/expo/modules/maps/googleMaps/GoogleMapsHeatmaps.kt]:
```suggestion
      tileOverlay?.let { heatmapOverlays.add(it) }
```

File: expo-maps/android/src/main/java/expo/modules/maps/googleMaps/GoogleMapsHeatmaps.kt [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+import com.google.android.gms.maps.model.TileOverlay
+import com.google.android.gms.maps.model.TileOverlayOptions
+import com.google.maps.android.data.kml.KmlLayer
+import com.google.maps.android.heatmaps.Gradient
+import com.google.maps.android.heatmaps.HeatmapTileProvider
+import com.google.maps.android.heatmaps.WeightedLatLng
+import expo.modules.maps.HeatmapObject
+import expo.modules.maps.interfaces.Heatmaps
+
+class GoogleMapsHeatmaps(private val context: Context, private val map: GoogleMap) : Heatmaps {
```
Comment (p-karcz) [inline expo-maps/android/src/main/java/expo/modules/maps/googleMaps/GoogleMapsHeatmaps.kt]: `context` is not used and can be removed

File: expo-maps/android/src/main/java/expo/modules/maps/PropsData.kt [hunk 9]
```diff
@@ -39,6 +40,11 @@ data class MarkerObject(
 }
 
 data class Point(@Field val latitude: Double = 0.0, @Field val longitude: Double = 0.0) : Record
```
Comment (p-karcz) [inline expo-maps/android/src/main/java/expo/modules/maps/PropsData.kt]:
```suggestion
data class Point(
  @Field val latitude: Double = 0.0,
  @Field val longitude: Double = 0.0
) : Record
```

**Pull request merged**